docs/stuck-ingests.md

```diff
@@ -0,0 +1,130 @@
+
+1. Reduce the high-water mark consumption of the ingest process for these files, so that it's within our RAM limit for the Sidekiq workers. This would be done by trying out some small optimisations (e.g. instead of an in-memory array of many `SubmissionEntry` instances, keep an array of `Hash` instances instead.
+2. Increase the RAM on the Sidekiq workers. I do not yet have an idea of how much RAM is _needed_ in order for these files to succeed.
+
+## Recovery from the problem
+
+Although we want to [prevent the out-of-memory errors](#prevention-of-out-of-memory-errors), we should still be able to recover from the case where it occurs. This is for two reasons:
+
+1. We _already_ have some stuck ingests that we need to fix (possibly manually).
+2. It will be difficult to guarantee that we have fixed the out-of-memory errors, and it's the sort of error which is quite easy to re-introduce as a regresssion.
+
+### What should the recovery process look like when viewed from outside?
+
+Since this failure is not a result of supplier user error, I think that the recovery should occur without the supplier user being made aware of it, and (hence) without requiring any supplier user intervention (that is, e.g. they shouldn't have to re-submit their return).
+
+**Before attempting any recovery (manual or automatic), we should confirm this with CCS.**
+
+### We want automatic recovery
+
+The ideal behaviour would be that after Sidekiq restarts, it automatically re-runs the killed job from the beginning. However, these links from Sidekiq's documentation make it clear that this is not the default behaviour of Sidekiq, and that we would have to upgrade to Sidekiq Pro to get it:
+
+* ["What happens to long-running jobs when Sidekiq restarts?"](https://github.com/mperham/sidekiq/wiki/FAQ#what-happens-to-long-running-jobs-when-sidekiq-restarts)
+* ["every time the Sidekiq process crashes, any messages being processed are lost. You can avoid this with Sidekiq Pro's reliable fetch feature."](https://github.com/mperham/sidekiq/wiki/Problems-and-Troubleshooting#my-sidekiq-process-is-crashing-what-do-i-do)
+* ["Sidekiq uses BRPOP to fetch a job from the queue in Redis. This is very efficient and simple but it has one drawback: the job is now removed from Redis. If Sidekiq crashes while processing that job, it is lost forever."](https://github.com/mperham/sidekiq/wiki/Reliability#using-super_fetch)
+
+### Is it safe to re-run `SubmissionIngestionJob`?
+
+I am not sure. I think that it is safe to re-run _as long as there are no other instances of this running for the same `SubmissionFile`_. Needs further investigation.
+
+### Fixing an individual stuck return
+
+__If__ it is [safe to rerun the job](#is-it-safe-to-re-run-submissioningestionjob`), then we can do exactly that:
+
+```ruby
+# In a Rails console
+SubmissionIngestionJob.perform_later(SubmissionFile.find(<ID>))
+```
+
+## What to do next
```

Comment:
We may also wish to consider paying for Sidekiq Pro in order that users do not need to retry their submissions whilst the investigation continues. I've not used it before and am not sure how costly it is to switch over and if there are other consequences of moving but it seems like a valuable piece of reliability. I'm imagining other situations where the sidekiq containers are deployed without connection draining and their long process may be terminated early.

Reply:
I've added it to the list, thanks!